Qiskit/saiba
# Store user progression in backend

## Changes

Fix #189 

This PR allows to store the progression of a user in a specific unit.

**Known issues**:
- https://github.com/Qiskit/saiba/issues/295
- When you log-out and log-in again the progression is not visible correcty in the UI.

## Implementation details

For this feature we created:
- Three services to: create a progression with its goal, create a goal and update a goal.
- Updated the method that sets the progression of a unit in the store.
- Updated the types of the different entities.

## How to read this PR

I would recommend to start the review from the `setUnitProgressionGoalFromWidget` method in `store\user\store.ts` and navigate through it.

In that method we just check for two things mainly:
- If the progression doesn't exist we create it.
- If the progression exists:
  - We create or update the goal.

All these actions are being supported for authenticated and non-authenticated users. The main difference between the data managed by these users are the `id` fields in those entities. For non authenticated users the progression and the goal don't have ids.

## How to test it

1. Sign-in in the application with your non-developer user: [preview](https://saiba-pr-294.tfb1j0wd7x2.us-south.codeengine.appdomain.cloud/)
2. Go to the `unit-03`
3. Just play with the two first widgets
4. You can check in the admin panel with your developer user the progression in your another user
